packages/dataviews/src/dataform-controls/array.tsx

```diff
@@ -18,68 +22,183 @@ export default function ArrayControl< Item >( {
-			// Convert TokenItem objects to strings
-			const stringTokens = tokens.map( ( token ) => {
-				if ( typeof token !== 'string' ) {
+	const validateTokens = useCallback(
+		( tokens: ( string | { value: string; label?: string } )[] ) => {
+			// Extract actual values from tokens for validation
+			const tokenValues = tokens.map( ( token ) => {
+				if ( typeof token === 'object' && 'value' in token ) {
 					return token.value;
 				}
+				return token;
+			} );
+
+			// First, check if elements validation is required and any tokens are invalid
+			if ( field.isValid?.elements && elements ) {
+				const invalidTokens = tokenValues.filter( ( tokenValue ) => {
+					return ! elements.some(
+						( element ) => element.value === tokenValue
+					);
+				} );
+
+				if ( invalidTokens.length > 0 ) {
+					setCustomValidity( {
+						type: 'invalid',
+						message: __(
+							'Please select from the available options.'
+						),
+					} );
+					return;
+				}
+			}
+
+			// Then check custom validation if provided
+			if ( field.isValid?.custom ) {
+				const result = field.isValid.custom(
+					{
+						...data,
+						[ id ]: tokenValues,
+					},
+					field
+				);
+
+				if ( result ) {
+					setCustomValidity( {
+						type: 'invalid',
+						message: result,
+					} );
+					return;
+				}
+			}
 
-				const tokenByLabel = findElementByLabel( token );
+			// If no validation errors, clear custom validity
+			setCustomValidity( undefined );
+		},
+		[ elements, data, id, field ]
+	);
 
-				return tokenByLabel?.value || token;
+	// Generate custom invalid message for elements validation
+	const customInvalidMessage = useMemo( () => {
+		if ( field.isValid?.elements && elements ) {
+			return __( 'Please select from the available options.' );
+		}
+		return __( 'Invalid item' );
+	}, [ field.isValid?.elements, elements ] );
+
+	const onChangeControl = useCallback(
+		( tokens: ( string | { value: string; label?: string } )[] ) => {
+			const valueTokens = tokens.map( ( token ) => {
+				if ( typeof token === 'object' && 'value' in token ) {
+					return token.value;
+				}
+				// If it's a string, it's either a new suggestion value or user input
+				return token;
 			} );
 
 			onChange( {
-				[ id ]: stringTokens,
+				[ id ]: valueTokens,
 			} );
+
+			validateTokens( tokens );
 		},
-		[ id, onChange, findElementByLabel ]
+		[ id, onChange, validateTokens ]
 	);
 
+	const onInputChange = useCallback( ( input: string ) => {
+		if ( input === '' ) {
+			// Reset custom validity when input is cleared
+			setCustomValidity( undefined );
+		}
+	}, [] );
+
 	return (
-		<FormTokenField
+		<ValidatedFormTokenField
+			required={ !! field.isValid?.required }
+			onValidate={ validateTokens }
+			customValidity={ customValidity }
 			label={ hideLabelFromVision ? undefined : label }
-			value={ arrayValue }
+			value={ arrayValueAsElements }
 			onChange={ onChangeControl }
+			onInputChange={ onInputChange }
```

Comment:
So, `onValidate` triggers for every `onChange` event. And the user is never able to introduce invalid data as per `__experimentalValidateInput`. So, the only case in which we ever display an error for `isValid.elements` is when the case in which the component loads with bad data (as [here](https://github.com/WordPress/gutenberg/pull/71194/files#r2354928714)). Is this correct?

If so, why we need this? The custom error for elements will never happen as a result of an input event, so it doesn't need to be cleared when input is empty. Perhaps I'm not considering another use case?

Reply:
The only reason for this is to catch this edge case, where the input is cleared but the message stays:

https://github.com/user-attachments/assets/25c63133-9f16-4fe6-8330-8f1b16b6d796

Reply:
@elazzabi I've pushed https://github.com/WordPress/gutenberg/pull/71194/commits/20410e48e6e9995bf14cd1982f6361599b984c72 to illustrate my point. It essentially removes everything related to input. We use `__experimentalValidateInput` to prevent an invalid element from being committed, but we don't trigger any validation on input.

I need to look at other things now, but there's this question we need to answer: does it makes sense to just remove `onValidate` / `customValidity` logic altogether? I think the only case where this may be useful is when the component is loaded with data that is not part of the element list.